### DISCUSSION TITLE
Validate all specific elements in JSON using Array Testing (Example: json[*].element)

### DISCUSSION THREAD

**Discussion by Artem0ne** [thunderclient/thunder-client-support#1278] [General]
Hello, 

we have question regards the JSON schema validation using array filters.
The JSON has the following structure:

``` 
[
{
"id" : 1
"status" :  "open"
},
{
"id" : 2
"status" :  "open"
},
{
"id" : 3
"status" :  "open"
}
]
```

In the area Test we use the following construction to validate the element **status**:

json[*].status  --> equal --> open

![image](https://github.com/rangav/thunder-client-support/assets/59470703/84d085f9-28f8-4964-b04d-978142782be3)
 
and this gives the error..
How can we check that **all** elements in the JSON have status "open"?

Small tip: the variable assignation _json[0].status_ (green marked on the picture) works fine.

**Comment by rangav** [comment]
Hi @Artem0ne 

please try this `json[status!=open]` -> Count -> 0

**Reply by Artem0ne** [reply]
hi @rangav 

Thank you for your response. I have a question regarding testing the array.

How can I test the array to ensure that only **these values** are considered valid within the JSON: ["open", "closed"]? We utilize filters in the URL to verify that only these specific values are present in the JSON.

I appreciate your prompt reply!

**Reply by rangav** [reply]
Do you know what other values is possible for the status?

- I can think of `json[status=pending]` count -> 0

**Reply by Artem0ne** [reply]
We have many objects with different types of statuses. For example ["open", "closed", "on hold", "active"] with the potential for further expansion. The resulting JSON varies based on the filter parameters specified within the URL.

Example 1:
**URL: {host}?status=open,closed**
Output:
```
[
{
"id" : 1
"status" :  "open"
},
{
"id" : 2
"status" :  "closed"
},
"id" : 3
"status" :  "open"
}
]
```
Example 2: 
**URL: {host}?status=active**
Output:
```
[
{
"id" : 5
"status" :  "active"
},
{
"id" : 6
"status" :  "active"
}
]
```

The question is how to test that **only the specific statuses** (as in **example 1** --> ?status=open,closed) that appear in JSON output.
It means that the status can be either "open" or "closed", but not others.  If status "active" appears in JSON, it must trigger an error.

Schema validation in Postman would be like:

```
 "properties" : {
                "status" : { "type" : "string" , "enum": ["open", "closed"]}, 
                }
```

**Reply by rangav** [reply]
Hi @Artem0ne You can do schema validation in Thunder Client also.

- You can use the **Tests** tab and select **scheme** action
- Or you can use **scripting** and manually perform schema validation and the function in `Post Request filter`
- Or you use scripting to write your custom logic for tests

Custom Scripting:
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#custom-filters

**Reply by Artem0ne** [reply]
hi @rangav,

we wanted to use this mechanism: **json[*].status --> equal --> open _or_ closed** instead of schema validation. 
Otherwise, we get too many different schemas as files and we want to avoid it. 

Is there any similar solution like **json[*].status --> equal --> open _or_ closed** to test 1 or many values?

**Reply by rangav** [reply]
Here are all possible ways you can test arrays
https://github.com/rangav/thunder-client-support/blob/master/docs/testing.md#test-arrays

Or You can also use built-in filters
https://github.com/rangav/thunder-client-support/blob/master/docs/filters.md#built-in-filters


If you have any suggestions for new built-in filters, we can add them

**Reply by Artem0ne** [reply]
The possible found solution for this:

**json[*]|filter(status!=open ) → count → 0**
